finetrainers/trainer.py

```diff
@@ -679,43 +711,68 @@ def train(self) -> None:
                             if "pooled_prompt_embeds" in text_conditions:
                                 text_conditions["pooled_prompt_embeds"].fill_(0)
 
-                    # These weighting schemes use a uniform timestep sampling and instead post-weight the loss
-                    weights = compute_density_for_timestep_sampling(
-                        weighting_scheme=self.args.flow_weighting_scheme,
-                        batch_size=batch_size,
-                        logit_mean=self.args.flow_logit_mean,
-                        logit_std=self.args.flow_logit_std,
-                        mode_scale=self.args.flow_mode_scale,
-                    )
-                    indices = (weights * self.scheduler.config.num_train_timesteps).long()
-                    sigmas = scheduler_sigmas[indices]
-                    timesteps = (sigmas * 1000.0).long()
-
-                    noise = torch.randn(
-                        latent_conditions["latents"].shape,
-                        generator=generator,
-                        device=accelerator.device,
-                        dtype=weight_dtype,
-                    )
-                    sigmas = expand_tensor_to_dims(sigmas, ndim=latent_conditions["latents"].ndim)
-                    noisy_latents = (1.0 - sigmas) * latent_conditions["latents"] + sigmas * noise
+                    is_flow = "Flow" in self.scheduler.__class__.__name__
+                    if is_flow:
+                        sigmas = self._derive_sigmas_for_flow(batch_size=batch_size, scheduler_sigmas=scheduler_sigmas)
+                        timesteps = (sigmas * 1000.0).long()
+                        noise, noisy_latents = self._calculate_noisy_latents_for_flow(
+                            sigmas=sigmas,
+                            latent_conditions=latent_conditions,
+                            generator=torch.Generator(self.state.accelerator.device).manual_seed(self.args.seed),
+                            weight_dtype=weight_dtype,
+                        )
+                    elif is_cog:
+                        timesteps = torch.randint(
```

Comment:
Can factor out into a different utility, too but let's settle on the current structure first.